instantiate plugin manager with per-restore logger so plugin logs are captured

skriss opened the pull request:
Fixes #1328
Replaces #1329 

I extracted the per-restore log into its own simple struct which makes fixing this a little bit easier, without having to combine the two methods mentioned in #1329. I think this is a little bit cleaner for now. Definitely want to further revisit this code (along with all of `pkg/restore`) post-v1.0.

cc @nrb @carlisia

skriss commented:
Ran some basic manual testing on this and confirmed that logs from plugins now show up in the per-restore log, whereas before they didn't.

skriss commented:
I can look at doing something like that, though to be clear the comments are not describing a new situation as a result of this PR - it was already true that the log object could be written to after being uploaded, and those log statements would not go into the uploaded log file.

nrb commented:
Yeah, that's true, but it still relies on programmers remembering that. Having an explicit close and throwing an error for write after close would, hopefully, catch any problems without finding out later that logging was being dropped unintentionally.

If it's not a huge value add, we can leave it out for now.

skriss commented:
Totally agree it's a good idea in general. Let me look at how much work it would be to accomplish in the current state. If we don't tackle it now, it would make sense to do when we do the rest of #286 to move all restore-related outputs into the log file. I have a feeling that there will be some more significant redesign as part of that, regardless.

skriss commented:
I did some refactoring in the latest commit to isolate usage of `restoreLog` to just a single function. I also renamed the restoreLog's `getReader(...)` function to `done(...)` and nil'ed out the `FieldLogger` when it's called, so any subsequent call to try to log with it will panic.  I also removed some useless code that was updating `restoreResult` objects in the case of restore failure, even though they'd never be uploaded.  I *hope* it's a little bit clearer now.  All of this code still needs a more thorough review/refactor post-1.0.

nrb commented:
I'll look over this again in the morning. Thanks for looking into closing out the log!

And agreed for a more holistic refactoring of the restore code in the future.

nrb commented:
LGTM - do you want to squash with an updated commit message?

skriss commented:
will do

skriss commented:
updated
